source/differentiation/ad/ad_helpers.cc

```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+          ADNumberTraits<ad_type>::is_tapeless == true)
+        {
+          Assert(
+            this->n_registered_independent_variables() ==
+              this->n_independent_variables(),
+            ExcMessage(
+              "Not all values of sensitivities have been registered or subsequently set!"));
+        }
+      Assert(this->n_registered_dependent_variables() ==
+               this->n_dependent_variables(),
+             ExcMessage("Not all dependent variables have been registered."));
+
+      // We can neglect correctly initializing the entries as
+      // we'll be overwriting them immediately in the succeeding call to
+      // Drivers::jacobian().
+      if (jacobian.m() != this->n_dependent_variables() ||
+          jacobian.n() != this->n_independent_variables())
+        jacobian.reinit({this->n_dependent_variables(),
+                         this->n_independent_variables()},
+                        true /*omit_default_initialization*/);
+
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(this->active_tape_index() !=
+                   Numbers<ad_type>::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+          Assert(this->is_recording() == false,
+                 ExcMessage(
+                   "Cannot compute Jacobian while tape is being recorded."));
+          Assert(this->independent_variable_values.size() ==
+                   this->n_independent_variables(),
+                 ExcDimensionMismatch(this->independent_variable_values.size(),
+                                      this->n_independent_variables()));
+
+          this->taped_driver.jacobian(this->active_tape_index(),
+                                      this->n_dependent_variables(),
+                                      this->independent_variable_values,
+                                      jacobian);
+        }
+      else
+        {
+          Assert(ADNumberTraits<ad_type>::is_tapeless == true,
+                 ExcInternalError());
+          Assert(this->independent_variables.size() ==
+                   this->n_independent_variables(),
+                 ExcDimensionMismatch(this->independent_variables.size(),
+                                      this->n_independent_variables()));
+
+          this->tapeless_driver.jacobian(this->independent_variables,
+                                         this->dependent_variables,
+                                         jacobian);
+        }
+
+      for (unsigned int i = 0; i < this->n_dependent_variables(); i++)
+        for (unsigned int j = 0; j < this->n_independent_variables(); j++)
+          // Because we perform just a single differentiation
+          // operation with respect to the "column" variables,
+          // we only need to consider them for symmetry conditions.
+          if (this->is_symmetric_independent_variable(j) == true)
+            jacobian[i][j] *= 0.5;
```

Comment:
Uh :-( That's the source of nightmares. Have you tested this on a simple example, preferably with a testcase?

You can make this code faster if you pull the loop over `i` all the way inside.

Reply:
Yes, I developed this and the counterparts in `ADHelperScalarFunction` using tests. These will all be added in the next PR. I've also been keeping my own repo in sync will all of these changes in order to make sure that none of them break the tests.